### DISCUSSION TITLE
How to export the orientations of the discrete ODF by certain numbers?

### DISCUSSION THREAD

**Discussion by louis96108** [mtex-toolbox/mtex#1890] [Ask Anything]
Hi,
I've used the following codes to obtain an ODF with different texture components, and discrete the ODF by 1000 random orientations.
```matlab
cs = crystalSymmetry('Fm-3m');
ss = specimenSymmetry('orthorhombic');
setMTEXpref('xAxisDirection','east');
setMTEXpref('zAxisDirection','outofPlane');
o = [orientation.goss(cs,ss) ...
    orientation.brass(cs,ss)...
    orientation.copper(cs,ss)];
c = [0.5,0.3,0.2];
odf = unimodalODF(o(:),'weights',c,'halfwidth', 10*degree);
h = [Miller(1,0,0,cs),Miller(1,1,0,cs),Miller(1,1,1,cs)];
o1 = odf.discreteSample(1000);
odf_rec = calcDensity(o1);
plot3d(odf_rec);
mtexColorbar('location','southOutSide','title','mrd');
figure
plotPDF(odf_rec,h,'antipodal','silent','upper','complete');
mtexColorbar('location','southOutSide','title','mrd');
```
![ODF](https://github.com/mtex-toolbox/mtex/assets/69557894/2547f2cc-d807-4d8c-a137-efb87102ace8)
The 3D odf looks good.
Then I tried to use the following codes to export the orientation of the discrete ODF into 2000 orientations.
```matlab
numOrientations = 2000;
orientations = calcOrientations(odf_rec, numOrientations);
export(orientations, 'orientations1.txt')
```
However, Matlab shows a warning: The command calcOrientations is depreciated! Please use discreteSamlpe instead.
How should I revise the codes? Looking forward to your reply. Thanks a lot.

**Comment by ralfHielscher** [comment]
In your first code you used `ori = odf.discreteSample(1000)` to compute the random orientations. In you second code you used the depreciated syntax `calcOrientations(odf_rec, numOrientations)`. Just replace `calcOrientations` by `discreteSample` which is exactly what the warning is telling you.

Ralf.

**Reply by louis96108** [reply]
Hi,
Thanks for your reply, I've exported the orientation by the method you told me. I used 'o1 = odf.discreteSample(1000)' to compute 1000 random orientation from the odf, and the whole number of the orientation I export is ‘2000’. So by using this method， can I get an odf that has 1000  random orientations and 1000 orientations containing several certain components, or in other words: 50% is random,50%  has texture?

**Reply by bright-men** [reply]
Hello, I would like to ask if there is a way to ensure that the result is the same every time when the discreteSample function is used to generate a specified number of orientations.

**Comment by ralfHielscher** [comment]
To add a 50 percent uniform portion do

```matlab
odf2  = 0.5*odf + 0.5
```

Ralf.

**Reply by louis96108** [reply]
Hi,
Thanks for your reply. As the last question shows, the basic 'odf' is defined by the following codes:

o = [orientation.goss(cs,ss) ...
    orientation.brass(cs,ss)...
    orientation.copper(cs,ss)];
c = [0.5,0.3,0.2];
odf = unimodalODF(o(:),'weights',c,'halfwidth', 10*degree);

And I've used the method that you told me: 'odf2=0.5*odf+0.5' to add 50% uniform and the 3D odf2 is shown like this: 
![0 5odf10 5](https://github.com/mtex-toolbox/mtex/assets/69557894/ba6dcf9c-dcc1-4a88-86ed-cc01313c7be2)
However, the 3D odf with out 50% uniform portion is shown like this:
![odf1](https://github.com/mtex-toolbox/mtex/assets/69557894/56c1c23b-c6a2-458d-9273-b026b54ee529)
The value of MRD has changed, and the method has induced random orientation.
So, What's the difference between ' odf2=0.5 odf + 0.5'  and 'discereSample'?  From 3D odf,  it looks like that 'discreteSample' connects spaces with random orientations.

**Reply by kilir** [reply]
Exactly, since you added a uniform portion to your odf, the density decreases. One of the differences to your example using discreteSample is that you estimate a new odf on your sample, hence if you sample a sufficiently large number of orientations, your result would approximate the input odf.

**Reply by louis96108** [reply]
OK, I'm a little confused about the concept of "uniform ODF "and the "random orientation", are they the same?
In my opinion, “uniform ODF” just influences the MRD
![织构1](https://github.com/mtex-toolbox/mtex/assets/69557894/bc4b78d0-bf72-4f08-ac18-c9983c01fd21)
BUT the “random orientation” compute by the 'discreteSample' will add some point in the write range of the Figure.
![织构1 - 副本](https://github.com/mtex-toolbox/mtex/assets/69557894/c536863b-e1ea-483d-8bfb-52ff83df1ccb)
Thank you so much.

**Reply by ddyyasfh** [reply]
Why do the orientations derived according to the above method not have corresponding weights?

**Reply by ralfHielscher** [reply]
It is a random sampling. A random sampling never has weights.
Ralf.

**Reply by louis96108** [reply]
@ddyyasfh You mean the weight of random orientation or the 'Goss', 'Brass', 'Copper' components?

**Reply by ralfHielscher** [reply]
Here is a one dimensional comparison with and without uniform portion. I hope this makes it more clear.
```
f = fibre.alpha(cs,ss)
plotFibre(odf2,f,'linewidth',2)
hold on
plotFibre(odf,f,'linewidth',2)
hold off
```
![odf](https://github.com/mtex-toolbox/mtex/assets/8619241/fa2cd295-fca9-48b3-82d5-a6b3445a4395)

Ralf.

**Reply by ddyyasfh** [reply]
For VPSC, it seems that the volume fraction corresponding to each direction of the fourth column is required
![image](https://github.com/mtex-toolbox/mtex/assets/104597723/a46b430b-279a-41af-a017-88e5a61d7e1c)

**Reply by ddyyasfh** [reply]
but I exported only the first three columns

**Reply by ddyyasfh** [reply]
`x=discreteSample(odf_rec,500)`
